test/unit/random_test.c

```diff
@@ -0,0 +1,228 @@
+/**
+ * Basic unit testing for random implementation, using Fourmilab's ent.
+ *
+ * Note that this is not a set of "true" randomness unit tests, the way
+ * DieHard is. It's intended to quickly expose pathological and trivial
+ * bugs and to test the API.
+ *
+ * We definitely should look into a true random unit test (TestU01,
+ * DieHard, DieHarder...)
+ */
+
+#include <runtime.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <string.h>
+
+/* Number of random bytes to output for test. */
+#define BSLEN 16384
+/* Number of bytes used by gen_bytestream's internal buffer */
+#define INTERNAL_BUFLEN 1024
+/* Number of values that ent gives out. */
+#define ENT_NSTATS 6
+/* Length of a stat entry key */
+#define KEY_LEN 64
+
+/* Maximum acceptable compressibility value for a reasonably long bytestream */
+#define COMPRESS_PASS_THRESH 10
+/* Minimum acceptable deviation from expected mean (127.5 in our case) */
+#define MEAN_DEV_THRESH 4
+
+struct ent_stat_entry {
+    char key[64];
+    double val;
+};
+
+/**
+ * Cleanup function: remove the bytestream file at path.
+ */
+static void rm_bytestream(char *path)
+{
+    remove(path);
+}
+
+/**
+ * Generate a bytestream of <len> random bytes in a file at path.
+ *
+ * TODO: buffer assumes len > 1024.
+ */
+static void gen_bytestream(char *path, int len)
+{
+    int fd, err;
+    u64 buf[INTERNAL_BUFLEN / 8];
+    u64 i, nbytes = 0;
+
+    fd = open(path, O_CREAT | O_WRONLY, S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP);
+    if (fd == -1)
+    {
+        msg_err("can't create bytestream file.\n");
+        exit(EXIT_FAILURE);
+    }
+
+    while (nbytes < BSLEN) {
+        for (i = 0; i < INTERNAL_BUFLEN / 8; i++)
+            buf[i] = random_u64();
+
+        nbytes += INTERNAL_BUFLEN;
+
+        err = write(fd, &buf[0], INTERNAL_BUFLEN);
+        if (err == -1)
+        {
+            msg_err("error writing to bytestream file.\n");
```

Comment:
Maybe print strerror(errno), and also check for EINTR here.

Reply:
Sure thing, thanks!